### PR TITLE
Realtime Reporting and S000, P000, R000

### PR DESCRIPTION
**Pull request by fedetony**
This is a PR using the bugfix 2.0.x branch

Added report and status change calls at the initial part of the Gcode functions and end part of the Gcode functions.
Sets the variable to the corresponding M_stategrbl value, and reports it if FULL_REPORT_TO_HOST_FEATURE is enabled.
For example when G0 is called: will set state to 5=Running Meanwhile it runs position will be handled by keep_alive function.
When Finishes the movement will synchronize and after the state will be set back to 3 = Idle.    This is important for handling the process from an external host. When the external host receives the Idle status, he can stream the next G-code command.

Since pause M0 and quick pause commands are different, the host can difference which pause command it has to send.
Notice M112 and M410 are still functional during quick pause feature.

### Requirements

Smile :). or none additional

### Description

Added reporting, pause-resume functionalities realtime as also auto-reporting state(TinyG grbl compatible) and position while moving.
Realtime commands are in the emergency parser next to M112 and M410.
Commands are:
P000 to quick pause
R000 to quick resume (Under pause state will only work M112,M410, and R000)
S000 to report position

For autoreporting:
if FULL_REPORT_TO_HOST_FEATURE the Machine state and position are reported while moving/homing/probing. State reports are GRBL style. and is just a number representing the state.
#0=init 1=reset, 2=alarm, 3=idle, 4=end, 5=run, 6=hold, 7=probe, 8=cycling, 9=homing, 10 =jogging 11=error
Note: Not all states are connected since they have to be modified in every gcode function called to set the state variable in the beginning and in the end.
For example when G0 is called: will set state to 5=Running Meanwhile it runs position will be handled by keep_alive function.
When Finishes the movement will synchronize and after the state will be set back to 3 = Idle.
This is very useful for handling the process from an external host. When the external host receives the Idle status, it can stream the next G-code command.
G0,G1,G2,G3,G27,G28,G29,G30,G33 are already altered.

`KEEPALIVE_INTERVAL_DIVIDER` divides the `KEEPALIVE_INTERVAL` for faster reporting than the actual 1 second.

Note: Some safety minimum rate must be set (like 200 ms) else there is a risk of blocking the serial buffers.

### Benefits
It works for CNC, Laser engraving, Drawing bots. Where is important to hold/pause the machine and have the possibility to resume the movement. This must react as fast as possible. Before this features, in marlin M0-M1 Pause commands were running after the command buffer and queue are executed. That could be catastrophic for a CNC machine.

### Configurations
Here The configuration of my machine:

[MarlinConfig.zip](https://github.com/MarlinFirmware/Marlin/files/5200474/MarlinConfig.zip)

for autoreporting:
You need: `HOST_KEEPALIVE_FEATURE`, `NANODLP_Z_SYNC` and `NANODLP_ALL_AXIS` for complete autoreporting

You need only to add to Configuration.h the two following lines (line 1499)
```cpp
#define KEEPALIVE_INTERVAL_DIVIDER 3 // Divide the KEEPALIVE_INTERVAL for faster reporting
#define FULL_REPORT_TO_HOST_FEATURE // Enable this to send Machine status reports while moving and status reports GRBL style
```

### Related Issues
#19253 
#17462 
#19324 <-the PR is to replace this one

For high inertia machines or while moving fast the machine may loose steps to pause abruptly. In my machine has not ever loose steps. yet deceleration tripping before calling quick_pause or quick _stop is possible.

### CODE REVIEW

**Comment by thinkyhead** [comment]
This needed a lot of fixing up. To get the fixed-up branch into your working copy:
```
git fetch origin
git checkout bugfix2.0.xautoreport
git reset --hard origin/bugfix2.0.xautoreport
```

**Comment by thinkyhead** [comment]
Note that suspending the stepper ISR in the middle of a move is never a good idea, so we limit it to use when killing a print job.  If the stepper was moving at 100mm/s when it stopped, it will try to jerk from zero to 100 instantly and will definitely lose position, so a proper closed loop is needed to allow for this.

Also, the suspension of the stepper ISR is only meant to be momentary. It cannot be suspended for any long duration and be guaranteed to stay suspended. Any new move (and various other commands, like `G92`) can cause the ISR to start up again, where it will continue processing the block that it was processing when it was suspended, causing the aforementioned loss of position.

**Comment by fedetony** [comment]
Great 👍 I'm very glad to share and contribute... and Thank you for all your efforts...

**Comment by mron** [comment]
I'm trying this out. S000 gives current position. It's also parsed as a command and give "echo:Unknown command: "S000"

**Comment by fedetony** [comment]
> I'm trying this out. S000 gives current position. It's also parsed as a command and give "echo:Unknown command: "S000"

Yes, Is because the normal gcode parser has not recognize the command. The instant command is handled in the emergency parser.  I did not modify it, but it can be done :) Easily. Is to include the S000, P000 and R000 commands in the ignore flag.

PS: The messages appear only after the command and queued commands execution.

**Comment by mron** [comment]
I don’t think the emergency parser is the place to handle this. It looks like a very fast parser to handle emergencies. Your codes require immediate attention, but aren’t emergencies. I’d suggest adding a new code prefix, IXXX or !XXX. These prefixes go to the parser and are handled immediately and not added to the queue.

**Comment by fedetony** [comment]
@mron There you go. Just change in the gcode folder the gcode.cpp and parser.cpp files :). The  parser is reading from a queue of commands, the emergency parser is reading the serial input. Its just a name, you could call it "serial parser". It was made for the emergency stop M112 command originally that is why the name is "emergency parser",

@thinkyhead Maybe you want to commit the last change :). Fixes the "Unkown command" message popping when using quick commands

**Comment by darwinmcgrath** [comment]
Hi all.
Thanks for looking into this.
I'm testing this because [cnc.js](https://cnc.js.org/) lost the ability to update axes info when I upgraded Marlin. The axes don't reflect current position data unless I send M114 manually. Position data is never updated during a run.

**Setup**:
Lowrider 2 (https://www.v1engineering.com/lowrider-cnc/)
Rambo 1.4
Marlin: https://github.com/V1EngineeringInc/MarlinBuilder/releases  (https://github.com/V1EngineeringInc/MarlinBuilder/releases/download/505/V1CNC_Rambo-2.0.6.1-src.zip)
Raspberry Pi running this image: https://docs.v1engineering.com/electronics/v1pi/

**After incorporating your code changes**
Manual Jogs:
Axes info is updated when I jog the machine using the console or the move buttons on the Axes widget (Yay!) but _1 step delayed._ (Dang!)
e.g.
Starting at X0 Y0
Send G0 Y1          Axes widget shows X 0 Y 0
Send G0 Y2          Axes widget shows X 0 Y 1
Send G0 X1          Axes widget shows X 0 Y 2
Send G0 X2          Axes widget shows X 1 Y 2

Also does this for the Z axis

Sending M114 updates Axes to correct values.

Running gcode script:
Running a really simple pocket set up as a surfacing job so straight lines in a rectangle.

Axes widget and Visualizer widget updating as code runs.
Still 1 step delayed from actual.

Questions:
To test the emergency pause do I just need to send P000? Console was updating with commands sent from cnc.js so wasn't sure how to inject P000.
What else can I do to help? Not a coder but willing to dig into this.

**Comment by mron** [comment]
Have you tried M114RD?

**Comment by darwinmcgrath** [comment]
@mron  Typing M114RD in the console returns the same info and does the same thing to the Axes widget data. M114 doesn't seem to have a problem. Returns accurate info when I send it. It's the updating of the coordinates in the Visualizer and Axes widgets that seem a jog behind.
Maybe this is a cnc.js thing. A mismatch between refreshes from what Fedetony's code posts and when cnc.js does a screen refresh? Otherwise, the code is sending data from the prior position.

**Comment by fedetony** [comment]
@darwinmcgrath thanks for testing it. What I added is the reporting while moving, yet you need to set active in your configuration_adv.h the NANODLP_Z_SYNC  and NANODLP_ALL_AXIS for it to show the position at the end of the movement too :). Movements as G0 G1 and others already had these implemented, reason why I adapted mine to only show only while moving.
Originally I wanted to use the same grbl commands (! , % and ?) but as I started digging into the  Marlin parser and found it would make conflicts with Prusa's commands and normally 1 character commands are very difficult to handle. 

I am not sure how you send your commands to the printer, you can use a terminal like Putty or teraterm and connect through the serial. or Use an interface like Universal Gcode sender or pronterface. Or have a Rpi with octoprint, also allows you to stream Gcode to your machine in realtime. Most software you get is possible to change the commands for certain actions in the configuration files just because in each interface are handled differently. I opted to build my own (it is work in progress, but is already functional with TinyG, grbl and Marlin interfaces), yet I am not sure what are your needs for the machine you have.  
Hope it helps :).

**Comment by darwinmcgrath** [comment]
@fedetony  That worked perfectly! 
I am using cncjs on a Rpi. It's not a printer though. It's a cnc machine.
cncjs is an http interface and when gcode is running the custom command options are greyed out and the console seems pretty busy. I can pause using the cncjs pause button but it's not a hard stop. The router continues until the current command is complete.
So I need to dig a little more to see how to implement your commands. P000 should hard stop in theory, correct? Assuming I can send the command real time.
Thanks!

**Comment by fedetony** [comment]
@darwinmcgrath I'm glad to hear that :)
Yes, In principle is to set in the cncjs the commands: 
`P000` for `!` , 
`S000` or `M114RD` for `?` , 
`R000` for `~` 
`M112` for `^X` 

the `!%` combination is `M410` to clear the buffer
Or I could modify them in Marlin for you, but they will not work in every machine and will not be inside Marlin's updates :( 
`P000` stops the movement immediately. From there you can hard stop `M112` (kills machine) , stop `M410` (resets buffer) , get position and state `S000` or resume `R000`.

**Comment by darwinmcgrath** [comment]
@fedetony Hi. It's been a little while but I had to revert to a stock version of Marlin.
After the changes above my cnc would stutter on certain pieces of gcode, pausing in between each G1 command. I think it's taking time to send location data. Cncjs was getting what appeared to be up-to-date coordinates, which was the idea.
The attached gifs show the difference between unchanged Marlin and modified. Sorry for the wobble but I think you can see the stutter in the second video.

It's running the following gcode (although I cut out the G0 in the video):

<details><summary>The G-Code</summary>

```gcode
G00 X129.5693 Y-1.7461 F780
G01 X122.6029 Y-1.7460 F780
G01 X115.7581 Y-1.7459 F780
G01 X109.0191 Y-1.7458 F780
G01 X102.3700 F780
G01 X95.7948 Y-1.7457 F780
G01 X89.2775 F780
G01 X82.8024 Y-1.7456 F780
G01 X76.3534 F780
G01 X69.9147 F780
G01 X63.4703 F780
G01 X57.0043 F780
G01 X50.5008 F780
G01 X43.9438 Y-1.7457 F780
G01 X37.3175 F780
G01 X30.6059 Y-1.7458 F780
G01 X23.7932 Y-1.7459 F780
G01 X16.8633 Y-1.7460 F780
G01 X9.8004 Y-1.7461 F780
G01 X2.5886 Y-1.7462 F780
```
</details>

Where should I start to investigate this? Is there additional info I should look at?
Thanks!

<details><summary>Some Animated Images</summary>

Smooth
Stutter
</details>

**Comment by fedetony** [comment]
@darwinmcgrath ... I see the issue. It can be due to several things:
1. Cncjs gcode streaming is waiting for Marlin's reply to send the new code and has delays. 
1s) Use the same code and run it from an SDcard directly. Then you can see if it is Marlin or cncjs the one causing the delays. If it runs fluently the issue is the cncjs (which is my primary suspect). 
2. Your controller HW for Marlin is having trouble to keep up. 
2s) For this you can try to reduce the reporting time: So `KEEPALIVE_INTERVAL_DIVIDER` set it to 2 and `KEEPALIVE_INTERVAL` to 1 for it to report every 500ms
reports are sent every `(KEEPALIVE_INTERVAL/KEEPALIVE_INTERVAL_DIVIDER)` in seconds
3) Communications are noisy and more info requires more effort: 
3s) You could try to decrease the Baudrate.

Try 1 first. And come back...

**Comment by darwinmcgrath** [comment]
@fedetony  1s)  Running the gcode from the sdcard results in the same stutter.
Will try 2s) next. Hopefully tonight.

**Comment by fedetony** [comment]
@darwinmcgrath check fix report # 10847 and # 16402. It may be caused by linear advance in your settings.

**Comment by darwinmcgrath** [comment]
@fedetony   2s) does not eliminate the stutter.
Here is the console output fwiw (console output using vanilla Marlin just shows "echo:busy: processing" 7 times):

<details><summary>Log Output</summary>

```
_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
S_XYZ:5
echo:busy: processing
S_XYZ:5
Z_move_comp
S_XYZ:3
Z_move_comp
echo:busy: processing
S_XYZ:3
```
</details>

**Comment by mron** [comment]
I think I have controlled stop/restart working. I'd like to incorporate that into this branch. How should I go about this?

**Comment by fedetony** [comment]
> I think I have controlled stop/restart working. I'd like to incorporate that into this branch. How should I go about this?

You could fork mine or fork the marlin original bugfix (better the original, make sure is the bugfix2.0.x). Then implement the required changes in the branch you created and using git commits so you and others can track the changes to the branch. Then upload them to your github account :) and then people can download and test the branch. If its functional send a pull request to Marlin. :)
If you fork mine I'll check your changes and if they work I'll merge them to my branch.

**Comment by mron** [comment]
I made a fork to my account. I made the changes using VSC. How do I push those changes back up to my GitHub copy? I thought committing the changes would do that.

**Comment by Vertabreak** [comment]
have to push it to its origin you only created a commit locally.

**Comment by mron** [comment]
It did work, I was looking in the wrong place

**Comment by mron** [comment]
Here's my fork: https://github.com/mron/Marlin/tree/bugfix2.0.xautoreport
It's a first cut. I need to test it more on my CNC. It seems to be working so far. I made changes to the queue, the planner, the stepper, and motion. Right now only x,y,z motion is working. I want to add spindle/laser. I'm not planning on dealing with extrusion. Only simple trapezoid speed ramps.
The commands P000,R000, and S000 are used to pause, resume and get status.
I'd like to incorporate M0/M1 and M108 to stop and restart.

**Comment by fedetony** [comment]
> Here's my fork: https://github.com/mron/Marlin/tree/bugfix2.0.xautoreport
> It's a first cut. I need to test it more on my CNC. It seems to be working so far. I made changes to the queue, the planner, the stepper, and motion. Right now only x,y,z motion is working. I want to add spindle/laser. I'm not planning on dealing with extrusion. Only simple trapezoid speed ramps.
> The commands P000,R000, and S000 are used to pause, resume and get status.
> I'd like to incorporate M0/M1 and M108 to stop and restart.

If you want to use a different code than P000, S000 or R000 go inside the emergency parser e_parser.h (not the cpp), weird, but is in the header implemented. You can use grbl commands, but be aware not to use a prusa printer ;). Or if you want to add M0/M1 and M108 that is the place. I'll check your changes, unfortunately I do not have a working machine in the moment for testing Marlin.

**Comment by fedetony** [comment]
@mron great work. :) Is very good idea to store the info until resumed. Also that you take in count the process SDcard/usb or serial to stop the feeds. I noticed yo liberate the memory on resume
Without testing and  just looking at the code, I believe is only missing:
- a feedhold_stop(); // Restore not continue movement function and liberate memory allocation or better release_feed_hold() to release_feed_hold_continue() and release_feed_hold_stop()
- other axis support and extruder. Just imagine you pause but it keeps extruding or the other axis keeps moving.
Some points to discuss:
-MINIMAL_STEP_RATE 10 // Do I need this? ( should be 120? )-> what units is it in? Difficult to say because depends on the machine. Yet could be added in the configuration. 
-Is maybe better to choose if to stop now by the size of the allocated block? Speeds and accelerations should not be used, since they depend on each machine.
-Also the movement has phases on the trapezoid: accelerate move in a speed and decelerate. There is no sense in stopping in the decelerate region. if the block is in deceleration phase, it should not be paused but allowed to finish? I think is doing this from the comments...
I'll buy a cheap SKR board to test this... :)

**Comment by mron** [comment]
Thanks for checking it out. Good points about needing to release the memory if the feedhold is abandoned. I can stop the extruder. I don't want to take on restarting print jobs. Insuring everything is heated and ready to go seems like a lot. I see this as a CNC/Laser function, I think printers are usually OK stopping at the end of a move.
I want to add a gpio response to trigger a pause.

**Comment by fedetony** [comment]
> take on restarting print jobs

The making sure Temps and all is ok for the extruder is already implemented in the program. You just need to trigger the resume, if it does not have the temperatures it keeps waiting and triggers the heating functions. It is relevant since some people may connect the PWM of the spindle/laser to the heater PWM.
But you need to trigger the pause/rersume in all steppers including the extruders and servos. The stop heating is triggered if you have it in the configuration set like that. I use the extruders 0  as 4th axis and E1 to drive my spindle.
The GPIO input is similar to setting the Spindle, for example I have in my SKR1.4 board Pins
`
//
// Spindle/Laser
//
#if EITHER(SPINDLE_FEATURE, LASER_FEATURE)
  #define SPINDLE_LASER_ENA_PIN E1_ENABLE_PIN   // digital pin
  #define SPINDLE_LASER_PWM_PIN    E1_STEP_PIN   // digital pin - MUST BE HARDWARE PWM
  #define SPINDLE_DIR_PIN          E1_DIR_PIN    // digital pin
#endif
//
//coolant
//
#if ENABLED(COOLANT_CONTROL)
  #define COOLANT_MIST_PIN E0_DIAG_PIN    // digital pin like endstop
  #define COOLANT_FLOOD_PIN E1_DIAG_PIN    // digital pin
#endif
`

I think you need to add the #if statement it to all boards for a general GPIO pause feature... though if is only yours, then add it to your board pinout like I did to my SKR1.4 to use it in a CNC/Laser/drawingbot in one machine.

**Comment by mron** [comment]
I'm pretty sure all the temp checking is done pre-block_buffering. Once a move is in the buffer, it's assumed it's good to go. I'll look into the wait-for-temp functions. It might not be as tricky as I fear.

**Comment by mron** [comment]
There is a "ensure_safe_temperature" in pause.cpp. Needs ADVANCE_PAUSE. Might work.
I'm thinking about when to abandon the feedhold. There are more G-codes and UI commands that should clear the condition than would use it. M24 and G28 should clear it.

**Comment by fedetony** [comment]
> There is a "ensure_safe_temperature" in pause.cpp. Needs ADVANCE_PAUSE. Might work.
> I'm thinking about when to abandon the feedhold. There are more G-codes and UI commands that should clear the condition than would use it. M24 and G28 should clear it.

When you make any command (pause command M0 for example), it goes to a queue, the parser selects the correspondent function to run for it following the queue FIFO. When moving the parser is disabled and when the movement is finished, then the parser makes the next queue command.
The emergency parser reads directly from the serial input, and will execute the instant commands. Yet you can call the same functions as the parser would. In the case of pause-stop, to call the same function M0 calls when stopped/resumed. The issue is that the stepper and gcode codes are disconnected in the program. So you need to create some intermediate  functions and include the *.h that call the correct process (same as the parser on M108 for resume). IMO It is better to reuse the code than duplicating or creating separate processes in the program, since changes in one are reflected to the other part of the program, they are already tested and maintained. Also all this temperature waiting and other stuff is connected to several machines/subprocesses and configurations... its going to be quite difficult to modify it or create it again.
Each gcode has its own .h and .cpp file, you need to include the specific heather and then you can call its functions.

**Comment by mron** [comment]
I think I'm going to leave all the temp check, parking etc to the host. I'll create a custom G-code command string that can be injected when the feedhold is complete.
How does the host know the P000 has completed? Is there an ACK sent? 
I added ! and ? commands. I suppressed the OK response for the ! and P in the gcode.cpp. This will be sent when feedhold is complete.

**Comment by fedetony** [comment]
> I think I'm going to leave all the temp check, parking etc to the host. I'll create a custom G-code command string that can be injected when the feedhold is complete.
> How does the host know the P000 has completed? Is there an ACK sent?

There is a state Variable. It is reported to the host. State reports are GRBL style. and is just a number representing the state.
#0=init 1=reset, 2=alarm, 3=idle, 4=end, 5=run, 6=hold, 7=probe, 8=cycling, 9=homing, 10 =jogging 11=error. It uses the same states as GRBL/tinyG machines.

**Comment by mron** [comment]
Ok, thanks, I'll look into this

**Comment by mron** [comment]
I have the ? and ! single letter commands working. I started to implement some of the grbl status codes. This is a big can of worms. I don't know how dedicated I am to that. I will work on spindle stop/Laser stop. That may be all for me.

**Comment by fedetony** [comment]
> I have the ? and ! single letter commands working. I started to implement some of the grbl status codes. This is a big can of worms. I don't know how dedicated I am to that. I will work on spindle stop/Laser stop. That may be all for me.

Yes, I used P000, S000 and R000 because they could be used by any machine. The ? command is used with prusa printers, and I would need to be selective with which machines inside the Feature. But you can put any, You can even put it to react without the \n as grbl does.

**Comment by mron** [comment]
I have the feedhole working with G2/G3 arc moves. Spindle/Laser stop/start is the last thing I'm adding. I have a version of Universal Gcode Sender working with Marlin as well.

**Comment by fedetony** [comment]
> I have the feedhole working with G2/G3 arc moves. Spindle/Laser stop/start is the last thing I'm adding. I have a version of Universal Gcode Sender working with Marlin as well.

Excellent, It looks pretty good to me. Got the new board to test this :), Some repairs in my DIY cnc and I'll concentrate on it. Commit your changes to your version.

**Comment by mron** [comment]
I need to merge them into my branch, it'll be up tomorrow

File: Marlin/src/module/planner.cpp [hunk 1]
```diff
@@ -1650,6 +1650,24 @@ void Planner::quick_stop() {
...
 }
 
+#if ENABLED(REALTIME_REPORTING_COMMANDS)
+
+  void Planner::quick_pause() {
+    // Suspend until quick_resume is called
+    // Don't empty buffers or queues
+    const bool did_suspend = stepper.suspend();
+    if (did_suspend)
+      TERN_(FULL_REPORT_TO_HOST_FEATURE, set_and_report_grblstate(M_HOLD));
```
Comment (thinkyhead) [inline Marlin/src/module/planner.cpp:1660]: I would enjoy hearing some audio samples of the horrific sounds the printer makes when you stop it dead in the middle of a fast move. These audio samples can be included as part of the documentation.

Comment (fedetony) [reply Marlin/src/module/planner.cpp:1660]: lol...
They would be exactly the same as when you press an end stop switch while moving fast, if the feature is active in your config. Uses the same stop mechanism.

**Comment by mron** [comment]
I've pushed the version with arc support. G2_G3 will smoothy stop and restart. I had to disable R000. It interfered with G2 RXXX commands. The eparser got triggered. Here's the link.
 
https://github.com/mron/Marlin/tree/bugfix2.0.xautoreport

I'll work on merging the changes from  @thinkyhead.

**Comment by mron** [comment]
I left some things in there that are specific to my set up. I had to change the SKR pins. And I enabled the spindle in the config files.

**Comment by fedetony** [comment]
> I've pushed the version with arc support. G2_G3 will smoothy stop and restart. I had to disable R000. It interfered with G2 RXXX commands. The eparser got triggered. Here's the link.
> 
> https://github.com/mron/Marlin/tree/bugfix2.0.xautoreport
> 
> I'll work on merging the changes from @thinkyhead.

Yes, they modified the parser.cpp and is broken. What did you modify exacly? I downloaded the new merged bugfix, but is giving me this error:
```
Marlin\src\gcode\parser.cpp: In static member function 'static void GCodeParser::parse(char*)':
Marlin\src\gcode\parser.cpp:235:17: error: duplicate case value
  235 |       case 'S': case 'P': case 'R': {
      |                 ^~~~
Marlin\src\gcode\parser.cpp:223:7: note: previously used here
  223 |       case 'P' ... 'Q':
      |       ^~~~
Marlin\src\gcode\parser.cpp:235:27: error: duplicate case value
  235 |       case 'S': case 'P': case 'R': {
      |                           ^~~~
Marlin\src\gcode\parser.cpp:220:27: note: previously used here
  220 |         case 'I' ... 'J': case 'R':
      |                           ^~~~
```
In my version I had S,R and P separated as the G commands. They were checked before the G command, and they did not interfere with G2.
I'll repair the parser.cpp to read the commands without conflicting with the G commands.

**Comment by mron** [comment]
Thanks.It didn’t make sense that it would interfere to me either.

**Comment by fedetony** [comment]
> And I enabled the spindle in the config files

hehehe, I did that too 👍 👍 I put these in the end of BTT_SKR_common.h file

```
//
// Spindle/Laser
//
#if EITHER(SPINDLE_FEATURE, LASER_FEATURE) // these pins are 3.3V
  #define SPINDLE_LASER_ENA_PIN    E1_ENABLE_PIN   // digital pin
  #define SPINDLE_LASER_PWM_PIN    E1_STEP_PIN   // digital pin - MUST BE HARDWARE PWM 
  #define SPINDLE_DIR_PIN          E1_DIR_PIN    // digital pin
#endif
//
//coolant
//
#if ENABLED(COOLANT_CONTROL) // these pins are 5V
  #define COOLANT_MIST_PIN         E0_DIAG_PIN    // digital pin like endstop
  #define COOLANT_FLOOD_PIN        E1_DIAG_PIN    // digital pin
#endif```

**Comment by fedetony** [comment]
This will fix the issue:
```
@@ -217,10 +217,10 @@ void GCodeParser::parse(char *p) {

     #if ENABLED(GCODE_MOTION_MODES)
       #if ENABLED(ARC_SUPPORT)
-        case 'I' ... 'J': case 'R':
+        case 'I' ... 'J':
           if (motion_mode_codenum != 2 && motion_mode_codenum != 3) return;
       #endif
-      case 'P' ... 'Q':
+      case 'Q':
         if (motion_mode_codenum != 5) return;
       case 'X' ... 'Z': case 'E' ... 'F':
         if (motion_mode_codenum < 0) return;
@@ -231,8 +231,37 @@ void GCodeParser::parse(char *p) {
       break;
     #endif // GCODE_MOTION_MODES

+    case 'R': {
+      #if ENABLED(GCODE_MOTION_MODES)
+        #if ENABLED(ARC_SUPPORT)
+          if (motion_mode_codenum != 2 && motion_mode_codenum != 3) return;
+        #endif
+      #endif
+      #if ENABLED(REALTIME_REPORTING_COMMANDS)
+        codenum = 0;                  // The only valid codenum is 0
+        uint8_t digits = 0;
+        while (*p++ == '0') digits++; // Count up '0' characters
+        command_letter = (digits == 3) ? letter : '?'; // Three '0' digits is a good command
+        return;
+      #endif
+    }
+
+    case 'P': {
+      #if ENABLED(GCODE_MOTION_MODES)
+        if (motion_mode_codenum != 5) return;
+      #endif
+      #if ENABLED(REALTIME_REPORTING_COMMANDS)
+        codenum = 0;                  // The only valid codenum is 0
+        uint8_t digits = 0;
+        while (*p++ == '0') digits++; // Count up '0' characters
+        command_letter = (digits == 3) ? letter : '?'; // Three '0' digits is a good command
+        return;
+      #endif
+
+    }
+
     #if ENABLED(REALTIME_REPORTING_COMMANDS)
-      case 'S': case 'P': case 'R': {
+      case 'S': {
         codenum = 0;                  // The only valid codenum is 0
         uint8_t digits = 0;
         while (*p++ == '0') digits++; // Count up '0' characters
~
~
```
can't make commits since it was merged :S

**Comment by mron** [comment]
I also modified M77. I added a couple lines to clear any feedhold data in order to cancel an active feedhold.

**Comment by fedetony** [comment]
@mron I think you need to make a new fork of the new bugfix, apply the changes you've made. You must have a fork version of the bugfix to be able to merge the changes in the future to the marlin.  Since my fork was merged, I can not make more changes to this fork and as consequence the changes you've made to add the soft stopping features are not in the merge. Probably you/we would need to start a new feature, improving the REALTIME_REPORTING_COMMANDS by adding soft stop as a feature. if you mention this pull request it will be commented and linked to your new feature.

**Comment by mron** [comment]
I'll give it a try

**Comment by mron** [comment]
Do I need to make a pull request? Or do I just make a fork from the current bugfix? I want to merge the soft stop/start code into the current bugfix.

**Comment by fedetony** [comment]
> Do I need to make a pull request? Or do I just make a fork from the current bugfix? I want to merge the soft stop/start code into the current bugfix.

https://jarv.is/notes/how-to-pull-request-fork-github/ this may help ;)

**Pull request merged**